spec/presenters/candidate_interface/personal_details_review_presenter_spec.rb

```diff
@@ -0,0 +1,219 @@
+          second_nationality: nil,
+        )
+
+        presenter = CandidateInterface::PersonalDetailsReviewPresenter.new(personal_details_form)
+
+        data = presenter.present
+
+        expect(data).to include(
+          key: t('application_form.personal_details.nationality.label'),
+          value: 'British',
+          action: t('application_form.personal_details.nationality.change_action'),
+        )
+      end
+
+      it 'includes a hash for dual nationalities' do
+        personal_details_form = FactoryBot.build(
+          :personal_details_form,
+          first_nationality: 'British',
+          second_nationality: 'Spanish',
+        )
+
+        presenter = CandidateInterface::PersonalDetailsReviewPresenter.new(personal_details_form)
+
+        data = presenter.present
+
+        expect(data).to include(
+          key: t('application_form.personal_details.nationality.label'),
+          value: 'British and Spanish',
+          action: t('application_form.personal_details.nationality.change_action'),
+        )
+      end
+    end
+
+    context 'when presenting English as the main language' do
+      it 'includes a hash with "Yes"' do
+        personal_details_form = FactoryBot.build(
+          :personal_details_form,
+          english_main_language: 'yes',
+          english_language_details: '',
+          other_language_details: '',
+        )
+
+        presenter = CandidateInterface::PersonalDetailsReviewPresenter.new(personal_details_form)
+
+        data = presenter.present
+
+        expect(data).to include(
+          key: t('application_form.personal_details.english_main_language.label'),
+          value: 'Yes',
+          action: t('application_form.personal_details.english_main_language.change_action'),
+        )
+      end
+
+      it 'excludes a hash for English language details if blank' do
+        personal_details_form = FactoryBot.build(
+          :personal_details_form,
+          english_main_language: 'yes',
+          english_language_details: '',
+          other_language_details: '',
+        )
+
+        presenter = CandidateInterface::PersonalDetailsReviewPresenter.new(personal_details_form)
+
+        data = presenter.present
+
+        expect(data).not_to include(
+          key: t('application_form.personal_details.english_main_language_details.label'),
+          value: '',
+          action: t('application_form.personal_details.english_main_language_details.change_action'),
+        )
+      end
+
+      it 'excludes a hash for other language details if given' do
+        personal_details_form = FactoryBot.build(
+          :personal_details_form,
+          english_main_language: 'yes',
+          english_language_details: '',
+          other_language_details: 'Broken? Oh man, are you cereal?',
+        )
+
+        presenter = CandidateInterface::PersonalDetailsReviewPresenter.new(personal_details_form)
+
+        data = presenter.present
+
+        expect(data).not_to include(
```

Comment:
Refactoring suggestion: instead of testing the entire hash, you could add a helper that checks that the language row of the field contains the correct value:

Perhaps something like:

```rb
expect(details_for(:other_language_details)).to eql("Broken? Oh man, are you cereal?")
```

This might clean up the assertions because you focus on the thing you want to test, and you can drop the test for the `action` key.

Reply:
Sweet, we've thought about this and done something similar:

Two helpers at the end of the file:

```ruby
  def present(form)
    CandidateInterface::PersonalDetailsReviewPresenter
      .new(form)
      .present
  end

  def row_for(key, value)
    {
      key: t("application_form.personal_details.#{key.to_s}.label"),
      value: value,
      action: t("application_form.personal_details.#{key.to_s}.change_action"),
    }
  end
```

Which allow us to write tests like this:

```ruby
      it 'includes a hash for other language details if given' do
        personal_details_form = FactoryBot.build(
          :personal_details_form,
          english_main_language: 'no',
          english_language_details: '',
          other_language_details: 'Broken? Oh man, are you cereal?',
        )

        expect(present(personal_details_form)).to include(
          row_for(:other_language_details, 'Broken? Oh man, are you cereal?')
        )
      end
```

We think this DRYs out the file a lot while still being readable (no `let`s), what do you think?